// @sendbird/uikit-react/ChannelSettings.js
import React__default from 'react';
import ChannelSettingsUI from './ChannelSettings/components/ChannelSettingsUI.js';
import { ChannelSettingsProvider } from './ChannelSettings/context.js';
import './useSendbirdStateContext.js';
import './withSendbird.js';
import './tslib.es6-c2fb729e.js';
import './ui/PlaceHolder.js';
import './LocalizationContext-287ee149.js';
import './stringSet-6422c6b9.js';
import './index-e4277f9c.js';
import './ui/Icon.js';
import 'prop-types';
import './index-92741c7e.js';
import './ui/Loader.js';
import './ui/IconButton.js';
import './ChannelSettings/components/ChannelProfile.js';
import './ui/ChannelAvatar.js';
import './ui/Avatar.js';
import './ui/ImageRenderer.js';
import './uuid-c15b8b3b.js';
import './utils-2c1f072b.js';
import './ui/TextButton.js';
import './color-57373b4c.js';
import './ChannelSettings/components/EditDetailsModal.js';
import './ui/Modal.js';
import 'react-dom';
import './utils-6c50a6d1.js';
import './index-09847c0f.js';
import './index-04bd1a59.js';
import './MediaQueryContext-114ca049.js';
import './ui/Button.js';
import './types-9d43ac6e.js';
import './ui/Input.js';
import './ChannelSettings/components/ModerationPanel.js';
import './ui/Accordion.js';
import './ui/AccordionGroup.js';
import './context-729493c4.js';
import './ui/Badge.js';
import './ui/Toggle.js';
import './ui/ContextMenu.js';
import './ui/SortByRow.js';
import './index-776554c0.js';
import './utils/message/getOutgoingMessageState.js';
import './ChannelSettings/components/UserListItem.js';
import './UserProfileContext-2001a468.js';
import './ui/MutedAvatarOverlay.js';
import './ui/UserProfile.js';
import './sendbirdSelectors.js';
import './pubSub/topics.js';
import './ui/UserListItem.js';
import './ui/Checkbox.js';
import '@sendbird/chat/groupChannel';
import './MemberList-4e71a841.js';
import './ChannelSettings/components/LeaveChannel.js';
import './ChannelSettings/components/UserPanel.js';

var ChannelSettings = function (props) {
  return /*#__PURE__*/React__default.createElement(ChannelSettingsProvider, {
    overrideInviteUser: props === null || props === void 0 ? void 0 : props.overrideInviteUser,
    channelUrl: props.channelUrl,
    onCloseClick: props === null || props === void 0 ? void 0 : props.onCloseClick,
    onLeaveChannel: props === null || props === void 0 ? void 0 : props.onLeaveChannel,
    onChannelModified: props === null || props === void 0 ? void 0 : props.onChannelModified,
    onBeforeUpdateChannel: props === null || props === void 0 ? void 0 : props.onBeforeUpdateChannel,
    queries: props === null || props === void 0 ? void 0 : props.queries,
    className: props === null || props === void 0 ? void 0 : props.className,
    disableUserProfile: props === null || props === void 0 ? void 0 : props.disableUserProfile,
    renderUserProfile: props === null || props === void 0 ? void 0 : props.renderUserProfile
  }, /*#__PURE__*/React__default.createElement(ChannelSettingsUI, {
    renderPlaceholderError: props === null || props === void 0 ? void 0 : props.renderPlaceholderError,
    renderChannelProfile: props === null || props === void 0 ? void 0 : props.renderChannelProfile,
    renderModerationPanel: props === null || props === void 0 ? void 0 : props.renderModerationPanel,
    renderLeaveChannel: props === null || props === void 0 ? void 0 : props.renderLeaveChannel
  }));
};

export { ChannelSettings as default };
