// @sendbird/uikit-react/ChannelList.js
import React__default from 'react';
import { C as ChannelListProvider } from './ChannelListProvider-eccd99c6.js';
import ChannelListUI from './ChannelList/components/ChannelListUI.js';
import './tslib.es6-c2fb729e.js';
import '@sendbird/chat/groupChannel';
import './pubSub/topics.js';
import './uuid-c15b8b3b.js';
import './utils-6c50a6d1.js';
import './consts-d9c5f25a.js';
import './UserProfileContext-2001a468.js';
import './useSendbirdStateContext.js';
import './withSendbird.js';
import './index-09847c0f.js';
import './index-776554c0.js';
import './utils/message/getOutgoingMessageState.js';
import './ChannelList/components/ChannelListHeader.js';
import './LocalizationContext-287ee149.js';
import './stringSet-6422c6b9.js';
import './index-e4277f9c.js';
import './index-92741c7e.js';
import 'prop-types';
import './ui/Avatar.js';
import './ui/ImageRenderer.js';
import './ui/Icon.js';
import './ChannelList/components/AddChannel.js';
import './ui/IconButton.js';
import './CreateChannel.js';
import './CreateChannel/components/CreateChannelUI.js';
import './CreateChannelProvider-a2885972.js';
import './sendbirdSelectors.js';
import './CreateChannel/components/InviteUsers.js';
import './ui/Modal.js';
import 'react-dom';
import './index-04bd1a59.js';
import './MediaQueryContext-114ca049.js';
import './ui/Button.js';
import './types-9d43ac6e.js';
import './ui/UserListItem.js';
import './ui/MutedAvatarOverlay.js';
import './ui/Checkbox.js';
import './ui/UserProfile.js';
import './ui/ContextMenu.js';
import './ui/SortByRow.js';
import './CreateChannel/components/SelectChannelType.js';
import './ChannelList/components/ChannelPreview.js';
import './_rollupPluginBabelHelpers-e92eb032.js';
import './ui/ChannelAvatar.js';
import './utils-2c1f072b.js';
import './ui/Badge.js';
import './index-4b4e9682.js';
import './index-2d27aeef.js';
import './ui/Loader.js';
import './index-f8a9cb51.js';
import './index-682a192c.js';
import './ui/MentionUserLabel.js';
import './consts-57c888b9.js';
import './ui/TextButton.js';
import './color-57373b4c.js';
import './Channel/components/TypingIndicator.js';
import './ChannelProvider-9ba5893a.js';
import './const-73db0277.js';
import './utils-aa4086dc.js';
import '@sendbird/chat/message';
import './useSendMultipleFilesMessage-b8256698.js';
import './compareIds-7e517404.js';
import './resolvedReplyType-fb987595.js';
import './useLongPress-f8655a14.js';
import './ChannelList/components/ChannelPreviewAction.js';
import './EditUserProfile.js';
import './EditUserProfile/context.js';
import './EditUserProfile/components/EditUserProfileUI.js';
import './ui/Input.js';
import './actionTypes-324495ed.js';
import './ui/PlaceHolder.js';

var ChannelList = function (props) {
  return /*#__PURE__*/React__default.createElement(ChannelListProvider, {
    className: props === null || props === void 0 ? void 0 : props.className,
    disableUserProfile: props === null || props === void 0 ? void 0 : props.disableUserProfile,
    allowProfileEdit: props === null || props === void 0 ? void 0 : props.allowProfileEdit,
    onBeforeCreateChannel: props === null || props === void 0 ? void 0 : props.onBeforeCreateChannel,
    onThemeChange: props === null || props === void 0 ? void 0 : props.onThemeChange,
    overrideInviteUser: props === null || props === void 0 ? void 0 : props.overrideInviteUser,
    onProfileEditSuccess: props === null || props === void 0 ? void 0 : props.onProfileEditSuccess,
    onChannelSelect: props === null || props === void 0 ? void 0 : props.onChannelSelect,
    sortChannelList: props === null || props === void 0 ? void 0 : props.sortChannelList,
    queries: props === null || props === void 0 ? void 0 : props.queries,
    disableAutoSelect: props === null || props === void 0 ? void 0 : props.disableAutoSelect,
    activeChannelUrl: props === null || props === void 0 ? void 0 : props.activeChannelUrl,
    isTypingIndicatorEnabled: props === null || props === void 0 ? void 0 : props.isTypingIndicatorEnabled,
    isMessageReceiptStatusEnabled: props === null || props === void 0 ? void 0 : props.isMessageReceiptStatusEnabled
  }, /*#__PURE__*/React__default.createElement(ChannelListUI, {
    renderChannelPreview: props === null || props === void 0 ? void 0 : props.renderChannelPreview,
    renderUserProfile: props === null || props === void 0 ? void 0 : props.renderUserProfile,
    renderHeader: props === null || props === void 0 ? void 0 : props.renderHeader,
    renderPlaceHolderEmptyList: props === null || props === void 0 ? void 0 : props.renderPlaceHolderEmptyList,
    renderPlaceHolderError: props === null || props === void 0 ? void 0 : props.renderPlaceHolderError,
    renderPlaceHolderLoading: props === null || props === void 0 ? void 0 : props.renderPlaceHolderLoading
  }));
};

export { ChannelList as default };
