// @sendbird/uikit-react/consts-d9c5f25a.js
var SCROLL_BUFFER = 10;
// voice message record
var VOICE_RECORDER_CLICK_BUFFER_TIME = 250;
var VOICE_RECORDER_DEFAULT_MIN = 1000; // 1 seconds
var VOICE_RECORDER_DEFAULT_MAX = 600000; // 10 minutes
var VOICE_RECORDER_AUDIO_BIT_RATE = 12000;
var VOICE_RECORDER_AUDIO_SAMPLE_RATE = 11025;
var BROWSER_SUPPORT_MIME_TYPE_LIST = ['audio/webm', 'audio/mp4', 'audio/mpeg', 'audio/ogg'];
/**
 * Append Audio element to the root
 * because I can't get the Audio element in the useEffect unmount scope
 */
var VOICE_PLAYER_ROOT_ID = 'sendbird-voice-player-provider-root';
var VOICE_PLAYER_AUDIO_ID = 'sendbird-global-audio-player-id';
// voice message file
var VOICE_MESSAGE_FILE_NAME = 'Voice_message.mp3';
var VOICE_MESSAGE_MIME_TYPE = 'audio/mp3;sbu_type=voice';
// meta array
var META_ARRAY_VOICE_DURATION_KEY = 'KEY_VOICE_MESSAGE_DURATION';
var META_ARRAY_MESSAGE_TYPE_KEY = 'KEY_INTERNAL_MESSAGE_TYPE';
var META_ARRAY_MESSAGE_TYPE_VALUE__VOICE = 'voice/mp3';
// delivery receipt in feature list
var DELIVERY_RECEIPT = 'delivery_receipt';
// file viewer slider
var SLIDER_BUTTON_ICON_SIDE_LENGTH = '32px';
// multiple files message file info count limit
var DEFAULT_MULTIPLE_FILES_MESSAGE_LIMIT = 10;

export { BROWSER_SUPPORT_MIME_TYPE_LIST as B, DEFAULT_MULTIPLE_FILES_MESSAGE_LIMIT as D, META_ARRAY_VOICE_DURATION_KEY as M, SCROLL_BUFFER as S, VOICE_RECORDER_DEFAULT_MAX as V, VOICE_RECORDER_DEFAULT_MIN as a, DELIVERY_RECEIPT as b, VOICE_MESSAGE_FILE_NAME as c, VOICE_MESSAGE_MIME_TYPE as d, META_ARRAY_MESSAGE_TYPE_KEY as e, META_ARRAY_MESSAGE_TYPE_VALUE__VOICE as f, VOICE_PLAYER_ROOT_ID as g, VOICE_PLAYER_AUDIO_ID as h, VOICE_RECORDER_AUDIO_BIT_RATE as i, VOICE_RECORDER_CLICK_BUFFER_TIME as j, SLIDER_BUTTON_ICON_SIDE_LENGTH as k, VOICE_RECORDER_AUDIO_SAMPLE_RATE as l };
