// @sendbird/uikit-react/CreateChannel.js
import React__default from 'react';
import CreateChannel$1 from './CreateChannel/components/CreateChannelUI.js';
import { C as CreateChannelProvider } from './CreateChannelProvider-a2885972.js';
import './CreateChannel/components/InviteUsers.js';
import './tslib.es6-c2fb729e.js';
import './LocalizationContext-287ee149.js';
import './stringSet-6422c6b9.js';
import './index-e4277f9c.js';
import './useSendbirdStateContext.js';
import './withSendbird.js';
import './ui/Modal.js';
import 'react-dom';
import './utils-6c50a6d1.js';
import './index-09847c0f.js';
import './index-04bd1a59.js';
import './MediaQueryContext-114ca049.js';
import './ui/IconButton.js';
import './ui/Button.js';
import './index-92741c7e.js';
import 'prop-types';
import './types-9d43ac6e.js';
import './ui/Icon.js';
import './ui/UserListItem.js';
import './UserProfileContext-2001a468.js';
import './ui/Avatar.js';
import './ui/ImageRenderer.js';
import './uuid-c15b8b3b.js';
import './ui/MutedAvatarOverlay.js';
import './ui/Checkbox.js';
import './ui/UserProfile.js';
import './sendbirdSelectors.js';
import './pubSub/topics.js';
import './ui/ContextMenu.js';
import './ui/SortByRow.js';
import './index-776554c0.js';
import './utils/message/getOutgoingMessageState.js';
import './CreateChannel/components/SelectChannelType.js';

var CreateChannel = function (props) {
  var onBeforeCreateChannel = props.onBeforeCreateChannel,
    userListQuery = props.userListQuery,
    onCreateChannel = props.onCreateChannel,
    overrideInviteUser = props.overrideInviteUser,
    onCancel = props.onCancel,
    renderStepOne = props.renderStepOne;
  return /*#__PURE__*/React__default.createElement(CreateChannelProvider, {
    onBeforeCreateChannel: onBeforeCreateChannel,
    userListQuery: userListQuery,
    onCreateChannel: onCreateChannel,
    overrideInviteUser: overrideInviteUser
  }, /*#__PURE__*/React__default.createElement(CreateChannel$1, {
    renderStepOne: renderStepOne,
    onCancel: onCancel
  }));
};

export { CreateChannel as default };
