// @sendbird/uikit-react/OpenChannelList.js
import React__default from 'react';
import OpenChannelListUI from './OpenChannelList/components/OpenChannelListUI.js';
import { O as OpenChannelListProvider } from './OpenChannelListProvider-45d4ab54.js';
import './OpenChannelList/components/OpenChannelPreview.js';
import './ui/Avatar.js';
import './tslib.es6-c2fb729e.js';
import './ui/ImageRenderer.js';
import './ui/Icon.js';
import 'prop-types';
import './uuid-c15b8b3b.js';
import './index-92741c7e.js';
import './stringSet-6422c6b9.js';
import './ui/PlaceHolder.js';
import './LocalizationContext-287ee149.js';
import './index-e4277f9c.js';
import './ui/Loader.js';
import './ui/IconButton.js';
import './CreateOpenChannel.js';
import './CreateOpenChannel/components/CreateOpenChannelUI.js';
import './types-9d43ac6e.js';
import './ui/Modal.js';
import 'react-dom';
import './utils-6c50a6d1.js';
import './index-09847c0f.js';
import './index-04bd1a59.js';
import './MediaQueryContext-114ca049.js';
import './ui/Button.js';
import './ui/Input.js';
import './ui/TextButton.js';
import './color-57373b4c.js';
import './CreateOpenChannel/context.js';
import './useSendbirdStateContext.js';
import './withSendbird.js';
import './consts-d9c5f25a.js';
import './pubSub/topics.js';

function OpenChannelList(_a) {
  var
    // provider
    className = _a.className,
    queries = _a.queries,
    onChannelSelected = _a.onChannelSelected,
    // ui
    renderHeader = _a.renderHeader,
    renderChannelPreview = _a.renderChannelPreview,
    renderPlaceHolderEmpty = _a.renderPlaceHolderEmpty,
    renderPlaceHolderError = _a.renderPlaceHolderError,
    renderPlaceHolderLoading = _a.renderPlaceHolderLoading;
  return /*#__PURE__*/React__default.createElement(OpenChannelListProvider, {
    className: className,
    queries: queries,
    onChannelSelected: onChannelSelected
  }, /*#__PURE__*/React__default.createElement(OpenChannelListUI, {
    renderHeader: renderHeader,
    renderChannelPreview: renderChannelPreview,
    renderPlaceHolderEmpty: renderPlaceHolderEmpty,
    renderPlaceHolderError: renderPlaceHolderError,
    renderPlaceHolderLoading: renderPlaceHolderLoading
  }));
}

export { OpenChannelList as default };
