// @sendbird/uikit-react/OpenChannelSettings.js
import React__default from 'react';
import OpenChannelUI from './OpenChannelSettings/components/OpenChannelSettingsUI.js';
import { OpenChannelSettingsProvider } from './OpenChannelSettings/context.js';
import './useSendbirdStateContext.js';
import './withSendbird.js';
import './tslib.es6-c2fb729e.js';
import './UserProfileContext-2001a468.js';
import './LocalizationContext-287ee149.js';
import './stringSet-6422c6b9.js';
import './index-e4277f9c.js';
import './index-92741c7e.js';
import 'prop-types';
import './ui/Icon.js';
import './ui/PlaceHolder.js';
import './ui/Loader.js';
import './OpenChannelSettings/components/OperatorUI.js';
import './OpenChannelSettings/components/OpenChannelProfile.js';
import './ui/TextButton.js';
import './color-57373b4c.js';
import './ui/OpenChannelAvatar.js';
import './ui/Avatar.js';
import './ui/ImageRenderer.js';
import './uuid-c15b8b3b.js';
import './utils-2c1f072b.js';
import './OpenChannelSettings/components/EditDetailsModal.js';
import './pubSub/topics.js';
import './ui/Modal.js';
import 'react-dom';
import './utils-6c50a6d1.js';
import './index-09847c0f.js';
import './index-04bd1a59.js';
import './MediaQueryContext-114ca049.js';
import './ui/IconButton.js';
import './ui/Button.js';
import './types-9d43ac6e.js';
import './ui/Input.js';
import './index-62d5f020.js';
import './ui/ContextMenu.js';
import './ui/SortByRow.js';
import './index-776554c0.js';
import './utils/message/getOutgoingMessageState.js';
import './context-729493c4.js';
import './ui/MutedAvatarOverlay.js';
import './ui/UserProfile.js';
import './sendbirdSelectors.js';
import '@sendbird/chat/openChannel';
import './ui/UserListItem.js';
import './ui/Checkbox.js';
import './ui/Accordion.js';
import './ui/AccordionGroup.js';

var OpenChannelSetting = function (props) {
  return /*#__PURE__*/React__default.createElement(OpenChannelSettingsProvider, {
    channelUrl: props === null || props === void 0 ? void 0 : props.channelUrl,
    onCloseClick: props === null || props === void 0 ? void 0 : props.onCloseClick,
    onBeforeUpdateChannel: props === null || props === void 0 ? void 0 : props.onBeforeUpdateChannel,
    onChannelModified: props === null || props === void 0 ? void 0 : props.onChannelModified,
    onDeleteChannel: props === null || props === void 0 ? void 0 : props.onDeleteChannel,
    disableUserProfile: props === null || props === void 0 ? void 0 : props.disableUserProfile,
    renderUserProfile: props === null || props === void 0 ? void 0 : props.renderUserProfile
  }, /*#__PURE__*/React__default.createElement(OpenChannelUI, {
    renderOperatorUI: props === null || props === void 0 ? void 0 : props.renderOperatorUI,
    renderParticipantList: props === null || props === void 0 ? void 0 : props.renderParticipantList
  }));
};

export { OpenChannelSetting as default };
