// @sendbird/uikit-react/RemoveMessageModal-007c2f04.js
import React__default, { useContext } from 'react';
import { Modal } from './ui/Modal.js';
import './tslib.es6-c2fb729e.js';
import './index-92741c7e.js';
import { B as ButtonTypes } from './types-9d43ac6e.js';
import { a as LocalizationContext } from './LocalizationContext-287ee149.js';
import { useThreadContext } from './Thread/context.js';
import { g as getModalDeleteMessageTitle } from './stringFormatterUtils-abbe70ba.js';

var RemoveMessage = function (props) {
  var _a;
  var onCancel = props.onCancel,
    onSubmit = props.onSubmit,
    message = props.message;
  var stringSet = useContext(LocalizationContext).stringSet;
  var deleteMessage = useThreadContext().deleteMessage;
  return /*#__PURE__*/React__default.createElement(Modal, {
    type: ButtonTypes.DANGER,
    disabled: ((_a = message === null || message === void 0 ? void 0 : message.threadInfo) === null || _a === void 0 ? void 0 : _a.replyCount) > 0,
    onCancel: onCancel,
    onSubmit: function () {
      deleteMessage(message).then(function () {
        onCancel === null || onCancel === void 0 ? void 0 : onCancel();
        onSubmit === null || onSubmit === void 0 ? void 0 : onSubmit();
      });
    },
    submitText: stringSet.MESSAGE_MENU__DELETE,
    titleText: getModalDeleteMessageTitle(stringSet, message)
  });
};

export { RemoveMessage as R };
