// @sendbird/uikit-react/Thread.js
import React__default from 'react';
import { ThreadProvider } from './Thread/context.js';
import ThreadUI from './Thread/components/ThreadUI.js';
import './useSendMultipleFilesMessage-b8256698.js';
import './pubSub/topics.js';
import './utils-aa4086dc.js';
import './tslib.es6-c2fb729e.js';
import './index-2d27aeef.js';
import './index-e4277f9c.js';
import './index-776554c0.js';
import './utils/message/getOutgoingMessageState.js';
import './index-09847c0f.js';
import '@sendbird/chat/message';
import './UserProfileContext-2001a468.js';
import './useSendbirdStateContext.js';
import './withSendbird.js';
import './Thread/context/types.js';
import '@sendbird/chat';
import '@sendbird/chat/groupChannel';
import './uuid-c15b8b3b.js';
import './consts-d9c5f25a.js';
import './LocalizationContext-287ee149.js';
import './stringSet-6422c6b9.js';
import './utils-fc1f5aa0.js';
import './index-92741c7e.js';
import 'prop-types';
import './Thread/components/ParentMessageInfo.js';
import './_rollupPluginBabelHelpers-e92eb032.js';
import './RemoveMessageModal-007c2f04.js';
import './ui/Modal.js';
import 'react-dom';
import './utils-6c50a6d1.js';
import './index-04bd1a59.js';
import './MediaQueryContext-114ca049.js';
import './ui/IconButton.js';
import './ui/Button.js';
import './types-9d43ac6e.js';
import './ui/Icon.js';
import './stringFormatterUtils-abbe70ba.js';
import './Thread/components/ParentMessageInfoItem.js';
import './ui/ImageRenderer.js';
import './ui/TextButton.js';
import './color-57373b4c.js';
import './ui/EmojiReactions.js';
import './ui/ReactionBadge.js';
import './ui/ReactionButton.js';
import './useLongPress-f8655a14.js';
import './ui/ContextMenu.js';
import './ui/SortByRow.js';
import './ui/BottomSheet.js';
import './ui/UserListItem.js';
import './ui/Avatar.js';
import './ui/MutedAvatarOverlay.js';
import './ui/Checkbox.js';
import './ui/UserProfile.js';
import './sendbirdSelectors.js';
import './ui/Tooltip.js';
import './ui/TooltipWrapper.js';
import './Message/context.js';
import './ui/VoiceMessageItemBody.js';
import './ui/ProgressBar.js';
import './VoicePlayer/useVoicePlayer.js';
import './index-e04159fe.js';
import './VoiceRecorder/context.js';
import './ui/PlaybackTime.js';
import './ui/Loader.js';
import './index-a0638563.js';
import './tokenize-eae44765.js';
import './ui/MentionLabel.js';
import './ui/LinkLabel.js';
import './useStatefulFileInfoList-5e68b93f.js';
import './index-0f017704.js';
import './ChannelProvider-9ba5893a.js';
import './const-73db0277.js';
import './compareIds-7e517404.js';
import './resolvedReplyType-fb987595.js';
import './Channel/components/SuggestedMentionList.js';
import './const-93731a4c.js';
import './ui/MessageItemMenu.js';
import './types-781e0bab.js';
import './ui/MessageItemReactionMenu.js';
import './ui/MessageInput.js';
import './index-6b3b1eca.js';
import 'dompurify';
import './consts-57c888b9.js';
import './consts-4ec7d640.js';
import './consts-be37687e.js';
import './index-c22e4772.js';
import './Message/hooks/useDirtyGetMentions.js';
import './Thread/components/ThreadHeader.js';
import './Thread/components/ThreadList.js';
import './Thread/components/ThreadListItem.js';
import './ui/DateSeparator.js';
import './index-4b4e9682.js';
import './index-f8a9cb51.js';
import './index-682a192c.js';
import './ui/TextMessageItemBody.js';
import './ui/OGMessageItemBody.js';
import './ui/FileMessageItemBody.js';
import './ui/ThumbnailMessageItemBody.js';
import './utils-bfc0ab3b.js';
import './ui/UnknownMessageItemBody.js';
import 'date-fns';
import './Thread/components/ThreadMessageInput.js';
import './useHandleUploadFiles-5ede1bfe.js';
import './VoiceRecorder/useVoiceRecorder.js';
import './index-493689bd.js';
import './hooks/useModal.js';
import './ui/PlaceHolder.js';

var Thread = function (props) {
  var
    // props
    className = props.className,
    // ThreadProviderProps
    channelUrl = props.channelUrl,
    message = props.message,
    onHeaderActionClick = props.onHeaderActionClick,
    onMoveToParentMessage = props.onMoveToParentMessage,
    isMultipleFilesMessageEnabled = props.isMultipleFilesMessageEnabled,
    // onBeforeSend~~~Message
    onBeforeSendUserMessage = props.onBeforeSendUserMessage,
    onBeforeSendFileMessage = props.onBeforeSendFileMessage,
    onBeforeSendVoiceMessage = props.onBeforeSendVoiceMessage,
    onBeforeSendMultipleFilesMessage = props.onBeforeSendMultipleFilesMessage,
    // ThreadUIProps
    renderHeader = props.renderHeader,
    renderParentMessageInfo = props.renderParentMessageInfo,
    renderMessage = props.renderMessage,
    renderMessageInput = props.renderMessageInput,
    renderCustomSeparator = props.renderCustomSeparator,
    renderParentMessageInfoPlaceholder = props.renderParentMessageInfoPlaceholder,
    renderThreadListPlaceHolder = props.renderThreadListPlaceHolder,
    renderFileUploadIcon = props.renderFileUploadIcon,
    renderVoiceMessageIcon = props.renderVoiceMessageIcon,
    renderSendMessageIcon = props.renderSendMessageIcon;
  return /*#__PURE__*/React__default.createElement("div", {
    className: "sendbird-thread ".concat(className)
  }, /*#__PURE__*/React__default.createElement(ThreadProvider, {
    channelUrl: channelUrl,
    message: message,
    onHeaderActionClick: onHeaderActionClick,
    onMoveToParentMessage: onMoveToParentMessage,
    onBeforeSendUserMessage: onBeforeSendUserMessage,
    onBeforeSendFileMessage: onBeforeSendFileMessage,
    onBeforeSendVoiceMessage: onBeforeSendVoiceMessage,
    onBeforeSendMultipleFilesMessage: onBeforeSendMultipleFilesMessage,
    isMultipleFilesMessageEnabled: isMultipleFilesMessageEnabled
  }, /*#__PURE__*/React__default.createElement(ThreadUI, {
    renderHeader: renderHeader,
    renderParentMessageInfo: renderParentMessageInfo,
    renderMessage: renderMessage,
    renderMessageInput: renderMessageInput,
    renderCustomSeparator: renderCustomSeparator,
    renderParentMessageInfoPlaceholder: renderParentMessageInfoPlaceholder,
    renderThreadListPlaceHolder: renderThreadListPlaceHolder,
    renderFileUploadIcon: renderFileUploadIcon,
    renderVoiceMessageIcon: renderVoiceMessageIcon,
    renderSendMessageIcon: renderSendMessageIcon
  })));
};

export { Thread as default };
