// @sendbird/uikit-react/cjs/ui/Modal.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var reactDom = require('react-dom');
var utils = require('../utils-eb6d90e0.js');
require('../tslib.es6-c74b513f.js');
require('../index-54279ec2.js');
var index = require('../index-b8ba4548.js');
var LocalizationContext = require('../LocalizationContext-58b6ed7c.js');
var MediaQueryContext = require('../MediaQueryContext-6141d955.js');
var ui_IconButton = require('./IconButton.js');
var ui_Button = require('./Button.js');
var ui_Icon = require('./Icon.js');
var ui_Label = require('../index-b5a9b4b7.js');
var types = require('../types-d0d770c3.js');
require('../stringSet-1539ac47.js');
require('../index-e8928da6.js');
require('prop-types');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var ModalHeader = function (_a) {
  var titleText = _a.titleText;
  return /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-modal__header"
  }, /*#__PURE__*/React__default["default"].createElement(ui_Label.Label, {
    type: ui_Label.LabelTypography.H_1,
    color: ui_Label.LabelColors.ONBACKGROUND_1
  }, titleText));
};
var ModalBody = function (_a) {
  var children = _a.children;
  return /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-modal__body"
  }, /*#__PURE__*/React__default["default"].createElement(ui_Label.Label, {
    type: ui_Label.LabelTypography.SUBTITLE_1,
    color: ui_Label.LabelColors.ONBACKGROUND_2
  }, children));
};
var ModalFooter = function (_a) {
  var submitText = _a.submitText,
    _b = _a.disabled,
    disabled = _b === void 0 ? false : _b,
    _c = _a.hideCancelButton,
    hideCancelButton = _c === void 0 ? false : _c,
    _d = _a.type,
    type = _d === void 0 ? types.ButtonTypes.DANGER : _d,
    onSubmit = _a.onSubmit,
    onCancel = _a.onCancel;
  var stringSet = React.useContext(LocalizationContext.LocalizationContext).stringSet;
  return /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-modal__footer"
  }, !hideCancelButton && /*#__PURE__*/React__default["default"].createElement(ui_Button["default"], {
    type: types.ButtonTypes.SECONDARY,
    onClick: onCancel
  }, /*#__PURE__*/React__default["default"].createElement(ui_Label.Label, {
    type: ui_Label.LabelTypography.BUTTON_1,
    color: ui_Label.LabelColors.ONBACKGROUND_1
  }, stringSet.BUTTON__CANCEL)), /*#__PURE__*/React__default["default"].createElement(ui_Button["default"], {
    type: type,
    disabled: disabled,
    onClick: onSubmit
  }, submitText));
};
function Modal(props) {
  var _a = props.children,
    children = _a === void 0 ? null : _a,
    _b = props.className,
    className = _b === void 0 ? '' : _b,
    _c = props.isCloseOnClickOutside,
    isCloseOnClickOutside = _c === void 0 ? false : _c,
    _d = props.isFullScreenOnMobile,
    isFullScreenOnMobile = _d === void 0 ? false : _d,
    titleText = props.titleText,
    submitText = props.submitText,
    _e = props.disabled,
    disabled = _e === void 0 ? false : _e,
    _f = props.hideFooter,
    hideFooter = _f === void 0 ? false : _f,
    _g = props.type,
    type = _g === void 0 ? types.ButtonTypes.DANGER : _g,
    /**
     * Do not use this! We will deprecate onCancel in v4.
     */
    _h = props.onCancel,
    /**
     * Do not use this! We will deprecate onCancel in v4.
     */
    onCancel = _h === void 0 ? utils.noop : _h,
    onClose = props.onClose,
    _j = props.onSubmit,
    onSubmit = _j === void 0 ? utils.noop : _j,
    renderHeader = props.renderHeader;
  var handleClose = onClose !== null && onClose !== void 0 ? onClose : onCancel;
  var isMobile = MediaQueryContext.useMediaQueryContext().isMobile;
  return /*#__PURE__*/reactDom.createPortal( /*#__PURE__*/React__default["default"].createElement("div", {
    className: "\n      sendbird-modal ".concat(className, "\n      ").concat(isFullScreenOnMobile && isMobile ? 'sendbird-modal--full-mobile' : '', "\n    ")
  }, /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-modal__content"
  }, (renderHeader === null || renderHeader === void 0 ? void 0 : renderHeader()) || /*#__PURE__*/React__default["default"].createElement(ModalHeader, {
    titleText: titleText !== null && titleText !== void 0 ? titleText : ''
  }), /*#__PURE__*/React__default["default"].createElement(ModalBody, null, children), !hideFooter && /*#__PURE__*/React__default["default"].createElement(ModalFooter, {
    disabled: disabled,
    onCancel: handleClose,
    onSubmit: onSubmit,
    submitText: submitText !== null && submitText !== void 0 ? submitText : '',
    type: type
  }), /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-modal__close"
  }, /*#__PURE__*/React__default["default"].createElement(ui_IconButton, {
    width: "32px",
    height: "32px",
    onClick: handleClose
  }, /*#__PURE__*/React__default["default"].createElement(ui_Icon["default"], {
    type: ui_Icon.IconTypes.CLOSE,
    fillColor: ui_Icon.IconColors.DEFAULT,
    width: "24px",
    height: "24px"
  })))), /*#__PURE__*/React__default["default"].createElement("div", {
    className: "\n          sendbird-modal__backdrop\n          ".concat(isCloseOnClickOutside && 'sendbird-modal__backdrop--clickoutside', "\n        "),
    onClick: function (e) {
      e === null || e === void 0 ? void 0 : e.stopPropagation();
      if (isCloseOnClickOutside) {
        handleClose();
      }
    }
  })), document.getElementById(index.MODAL_ROOT));
}

exports.Modal = Modal;
exports.ModalBody = ModalBody;
exports.ModalFooter = ModalFooter;
exports.ModalHeader = ModalHeader;
exports["default"] = Modal;
