// @sendbird/uikit-react/cjs/ui/QuoteMessageInput.js
'use strict';

var React = require('react');
var index = require('../index-dbe8dd84.js');
var ui_Icon = require('./Icon.js');
var ui_Label = require('../index-b5a9b4b7.js');
var LocalizationContext = require('../LocalizationContext-58b6ed7c.js');
var ui_ImageRenderer = require('./ImageRenderer.js');
var utils = require('../utils-93ca9d57.js');
require('../tslib.es6-c74b513f.js');
require('../utils/message/getOutgoingMessageState.js');
require('../index-54279ec2.js');
require('prop-types');
require('../stringSet-1539ac47.js');
require('../index-e8928da6.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var componentClassname = 'sendbird-quote_message_input__avatar';
function QuoteMessageThumbnail(_a) {
  var _b, _c;
  var message = _a.message;
  if (!index.isFileMessage(message) && !index.isMultipleFilesMessage(message) || index.isVoiceMessage(message)) {
    return null;
  }
  var thumbnailUrl = utils.getMessageFirstFileThumbnailUrl(message);
  if (!thumbnailUrl) {
    if (((_b = message.isFileMessage) === null || _b === void 0 ? void 0 : _b.call(message)) && (index.isImageMessage(message) || index.isVideoMessage(message))) {
      thumbnailUrl = utils.getMessageFirstFileUrl(message);
    } else if (((_c = message.isMultipleFilesMessage) === null || _c === void 0 ? void 0 : _c.call(message)) && message.fileInfoList.length > 0 && index.isImageFileInfo(message.fileInfoList[0])) {
      thumbnailUrl = message.fileInfoList[0].url;
    }
  }
  if (index.isVideoMessage(message) && thumbnailUrl) {
    return /*#__PURE__*/React__default["default"].createElement("div", {
      className: componentClassname
    }, /*#__PURE__*/React__default["default"].createElement("video", {
      style: {
        width: '44px',
        height: '44px'
      },
      src: thumbnailUrl
    }));
  } else if ((index.isThumbnailMessage(message) || index.isMultipleFilesMessage(message)) && thumbnailUrl) {
    return /*#__PURE__*/React__default["default"].createElement(ui_ImageRenderer["default"], {
      className: componentClassname,
      url: thumbnailUrl,
      alt: utils.getMessageFirstFileType(message),
      width: "44px",
      height: "44px",
      fixedSize: true
    });
  } else if (index.isAudioMessage(message)) {
    return /*#__PURE__*/React__default["default"].createElement("div", {
      className: componentClassname
    }, /*#__PURE__*/React__default["default"].createElement(ui_Icon["default"], {
      type: ui_Icon.IconTypes.FILE_AUDIO,
      fillColor: ui_Icon.IconColors.ON_BACKGROUND_2,
      width: "24px",
      height: "24px"
    }));
  } else {
    return /*#__PURE__*/React__default["default"].createElement("div", {
      className: componentClassname
    }, /*#__PURE__*/React__default["default"].createElement(ui_Icon["default"], {
      type: ui_Icon.IconTypes.FILE_DOCUMENT,
      fillColor: ui_Icon.IconColors.ON_BACKGROUND_2,
      width: "24px",
      height: "24px"
    }));
  }
}

function QuoteMessageInput(_a) {
  var className = _a.className,
    replyingMessage = _a.replyingMessage,
    onClose = _a.onClose;
  var stringSet = React.useContext(LocalizationContext.LocalizationContext).stringSet;
  var fileMessage = replyingMessage;
  var sender = replyingMessage === null || replyingMessage === void 0 ? void 0 : replyingMessage.sender;
  var displayFileIcon = (index.isFileMessage(replyingMessage) || index.isMultipleFilesMessage(replyingMessage)) && !index.isVoiceMessage(replyingMessage);
  return /*#__PURE__*/React__default["default"].createElement("div", {
    className: index.getClassName(['sendbird-quote_message_input', className])
  }, displayFileIcon && /*#__PURE__*/React__default["default"].createElement(QuoteMessageThumbnail, {
    message: fileMessage
  }), /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-quote_message_input__body",
    style: {
      width: "calc(100% - ".concat(displayFileIcon ? '164px' : '120px', ")"),
      left: displayFileIcon ? '92px' : '40px'
    }
  }, /*#__PURE__*/React__default["default"].createElement(ui_Label.Label, {
    className: "sendbird-quote_message_input__body__sender-name",
    type: ui_Label.LabelTypography.CAPTION_1,
    color: ui_Label.LabelColors.ONBACKGROUND_1
  }, "".concat(stringSet.QUOTE_MESSAGE_INPUT__REPLY_TO, " ").concat(sender && sender.nickname ? sender.nickname : stringSet.NO_NAME)), /*#__PURE__*/React__default["default"].createElement(ui_Label.Label, {
    className: "sendbird-quote_message_input__body__message-content",
    type: ui_Label.LabelTypography.BODY_2,
    color: ui_Label.LabelColors.ONBACKGROUND_3
  }, index.isImageMessage(fileMessage) && !index.isGifMessage(fileMessage) && stringSet.QUOTE_MESSAGE_INPUT__FILE_TYPE_IMAGE, index.isVideoMessage(fileMessage) && stringSet.QUOTE_MESSAGE_INPUT__FILE_TYPE__VIDEO, index.isGifMessage(fileMessage) && stringSet.QUOTE_MESSAGE_INPUT__FILE_TYPE_GIF, index.isUserMessage(replyingMessage) && replyingMessage.message, index.getUIKitMessageType(replyingMessage) === index.UIKitMessageTypes.FILE && utils.getMessageFirstFileName(fileMessage), index.isMultipleFilesMessage(replyingMessage) && "".concat(fileMessage.fileInfoList.length, " Photos"), index.isVoiceMessage(replyingMessage) && stringSet.VOICE_MESSAGE)), /*#__PURE__*/React__default["default"].createElement(ui_Icon["default"], {
    className: "sendbird-quote_message_input__close-button",
    type: ui_Icon.IconTypes.CLOSE,
    fillColor: ui_Icon.IconColors.ON_BACKGROUND_2,
    width: "24px",
    height: "24px",
    onClick: function () {
      return onClose(replyingMessage);
    }
  }));
}

module.exports = QuoteMessageInput;
