// @sendbird/uikit-tools/dist/cjs/uikitConfiguration/internal/initialConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initialConfig = void 0;
exports.initialConfig = {
    common: {
        enableUsingDefaultUserProfile: false,
    },
    groupChannel: {
        channel: {
            enableMention: false,
            enableOgtag: true,
            enableReactions: true,
            enableTypingIndicator: true,
            enableVoiceMessage: false,
            input: {
                camera: {
                    enablePhoto: true,
                    enableVideo: true,
                },
                enableDocument: true,
                gallery: {
                    enablePhoto: true,
                    enableVideo: true,
                },
            },
            replyType: 'quote_reply',
            threadReplySelectType: 'thread',
        },
        channelList: {
            enableMessageReceiptStatus: false,
            enableTypingIndicator: false,
        },
        setting: {
            enableMessageSearch: false,
        },
    },
    openChannel: {
        channel: {
            enableOgtag: true,
            input: {
                camera: {
                    enablePhoto: true,
                    enableVideo: true,
                },
                enableDocument: true,
                gallery: {
                    enablePhoto: true,
                    enableVideo: true,
                },
            },
        },
    },
};
