// @sendbird/uikit-react/cjs/Channel.js
'use strict';

var React = require('react');
var Channel_context = require('./ChannelProvider-fc111e5f.js');
var Channel_components_ChannelUI = require('./Channel/components/ChannelUI.js');
require('./UserProfileContext-c6d7abb7.js');
require('./useSendbirdStateContext.js');
require('./withSendbird.js');
require('./tslib.es6-c74b513f.js');
require('./const-c9e7a340.js');
require('./utils-9193a195.js');
require('./index-3448df64.js');
require('./index-e8928da6.js');
require('./index-dbe8dd84.js');
require('./utils/message/getOutgoingMessageState.js');
require('./index-54279ec2.js');
require('@sendbird/chat/message');
require('./useSendMultipleFilesMessage-d6a3d5cf.js');
require('./pubSub/topics.js');
require('./compareIds-28c6af62.js');
require('@sendbird/chat/groupChannel');
require('./uuid-9a117e16.js');
require('./_rollupPluginBabelHelpers-f5b0dfb9.js');
require('./consts-f54b15c2.js');
require('./resolvedReplyType-c4564ebe.js');
require('./ui/PlaceHolder.js');
require('./LocalizationContext-58b6ed7c.js');
require('./stringSet-1539ac47.js');
require('./ui/Icon.js');
require('prop-types');
require('./index-b5a9b4b7.js');
require('./ui/Loader.js');
require('./ui/ConnectionStatus.js');
require('./Channel/components/ChannelHeader.js');
require('./utils-e6c4b2b0.js');
require('./ui/IconButton.js');
require('./ui/ChannelAvatar.js');
require('./ui/Avatar.js');
require('./ui/ImageRenderer.js');
require('./utils-98441ba6.js');
require('./MediaQueryContext-6141d955.js');
require('./utils-eb6d90e0.js');
require('./Channel/components/MessageList.js');
require('./Channel/components/Message.js');
require('./Channel/components/SuggestedMentionList.js');
require('./const-876980e8.js');
require('./Thread/context.js');
require('./Thread/context/types.js');
require('@sendbird/chat');
require('./ui/DateSeparator.js');
require('./color-5bf5a8f5.js');
require('./ui/MessageInput.js');
require('./ui/Button.js');
require('./types-d0d770c3.js');
require('./index-cc2e05a4.js');
require('dompurify');
require('./consts-b3c4f548.js');
require('./consts-61d83828.js');
require('./consts-b711321c.js');
require('./tokenize-d5286fa8.js');
require('./ui/MessageContent.js');
require('./ui/UserProfile.js');
require('./sendbirdSelectors.js');
require('./index-520e12b8.js');
require('./index-7627b546.js');
require('./index-134e9852.js');
require('./ui/MessageItemMenu.js');
require('./ui/ContextMenu.js');
require('react-dom');
require('./ui/SortByRow.js');
require('./types-feda9656.js');
require('./ui/MessageItemReactionMenu.js');
require('./ui/ReactionButton.js');
require('./useLongPress-18b7af79.js');
require('./ui/EmojiReactions.js');
require('./ui/ReactionBadge.js');
require('./ui/BottomSheet.js');
require('./index-b8ba4548.js');
require('./ui/UserListItem.js');
require('./ui/MutedAvatarOverlay.js');
require('./ui/Checkbox.js');
require('./ui/Tooltip.js');
require('./ui/TooltipWrapper.js');
require('./Message/context.js');
require('./ui/AdminMessage.js');
require('./ui/TextMessageItemBody.js');
require('./index-ce07a2a3.js');
require('./ui/MentionLabel.js');
require('./ui/LinkLabel.js');
require('./ui/FileMessageItemBody.js');
require('./ui/TextButton.js');
require('./ui/ThumbnailMessageItemBody.js');
require('./utils-93ca9d57.js');
require('./ui/OGMessageItemBody.js');
require('./ui/UnknownMessageItemBody.js');
require('./ui/QuoteMessage.js');
require('./index-e5e6d9b0.js');
require('./ui/ThreadReplies.js');
require('./ui/VoiceMessageItemBody.js');
require('./ui/ProgressBar.js');
require('./VoicePlayer/useVoicePlayer.js');
require('./index-a95c4ab4.js');
require('./VoiceRecorder/context.js');
require('./ui/Modal.js');
require('./ui/PlaybackTime.js');
require('./useStatefulFileInfoList-c510792f.js');
require('./index-bc6623d7.js');
require('./Channel/components/FileViewer.js');
require('./Channel/components/RemoveMessageModal.js');
require('./stringFormatterUtils-92dbef3c.js');
require('./index-692ea6aa.js');
require('./Message/hooks/useDirtyGetMentions.js');
require('./Channel/utils/getMessagePartsInfo.js');
require('./Channel/utils/compareMessagesForGrouping.js');
require('./Channel/components/UnreadCount.js');
require('./Channel/components/FrozenNotification.js');
require('./Channel/components/TypingIndicator.js');
require('./Channel/components/MessageInput.js');
require('./ui/QuoteMessageInput.js');
require('./useHandleUploadFiles-39e8d107.js');
require('./VoiceRecorder/useVoiceRecorder.js');
require('./index-4c378655.js');
require('./hooks/useModal.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var Channel = function (props) {
  return /*#__PURE__*/React__default["default"].createElement(Channel_context.ChannelProvider, {
    channelUrl: props === null || props === void 0 ? void 0 : props.channelUrl,
    isReactionEnabled: props === null || props === void 0 ? void 0 : props.isReactionEnabled,
    isMessageGroupingEnabled: props === null || props === void 0 ? void 0 : props.isMessageGroupingEnabled,
    isMultipleFilesMessageEnabled: props === null || props === void 0 ? void 0 : props.isMultipleFilesMessageEnabled,
    showSearchIcon: props === null || props === void 0 ? void 0 : props.showSearchIcon,
    animatedMessage: props === null || props === void 0 ? void 0 : props.animatedMessage,
    highlightedMessage: props === null || props === void 0 ? void 0 : props.highlightedMessage,
    startingPoint: props === null || props === void 0 ? void 0 : props.startingPoint,
    onBeforeSendUserMessage: props === null || props === void 0 ? void 0 : props.onBeforeSendUserMessage,
    onBeforeSendFileMessage: props === null || props === void 0 ? void 0 : props.onBeforeSendFileMessage,
    onBeforeUpdateUserMessage: props === null || props === void 0 ? void 0 : props.onBeforeUpdateUserMessage,
    onBeforeSendVoiceMessage: props === null || props === void 0 ? void 0 : props.onBeforeSendVoiceMessage,
    onBeforeSendMultipleFilesMessage: props === null || props === void 0 ? void 0 : props.onBeforeSendMultipleFilesMessage,
    onChatHeaderActionClick: props === null || props === void 0 ? void 0 : props.onChatHeaderActionClick,
    onSearchClick: props === null || props === void 0 ? void 0 : props.onSearchClick,
    onBackClick: props === null || props === void 0 ? void 0 : props.onBackClick,
    replyType: props === null || props === void 0 ? void 0 : props.replyType,
    threadReplySelectType: props === null || props === void 0 ? void 0 : props.threadReplySelectType,
    queries: props === null || props === void 0 ? void 0 : props.queries,
    renderUserProfile: props === null || props === void 0 ? void 0 : props.renderUserProfile,
    filterMessageList: props === null || props === void 0 ? void 0 : props.filterMessageList,
    disableUserProfile: props === null || props === void 0 ? void 0 : props.disableUserProfile,
    disableMarkAsRead: props === null || props === void 0 ? void 0 : props.disableMarkAsRead,
    onReplyInThread: props === null || props === void 0 ? void 0 : props.onReplyInThread,
    onQuoteMessageClick: props === null || props === void 0 ? void 0 : props.onQuoteMessageClick,
    onMessageAnimated: props === null || props === void 0 ? void 0 : props.onMessageAnimated,
    onMessageHighlighted: props === null || props === void 0 ? void 0 : props.onMessageHighlighted,
    scrollBehavior: props.scrollBehavior,
    reconnectOnIdle: props.reconnectOnIdle
  }, /*#__PURE__*/React__default["default"].createElement(Channel_components_ChannelUI, {
    isLoading: props === null || props === void 0 ? void 0 : props.isLoading,
    renderPlaceholderLoader: props === null || props === void 0 ? void 0 : props.renderPlaceholderLoader,
    renderPlaceholderInvalid: props === null || props === void 0 ? void 0 : props.renderPlaceholderInvalid,
    renderPlaceholderEmpty: props === null || props === void 0 ? void 0 : props.renderPlaceholderEmpty,
    renderChannelHeader: props === null || props === void 0 ? void 0 : props.renderChannelHeader,
    renderMessage: props === null || props === void 0 ? void 0 : props.renderMessage,
    renderMessageInput: props === null || props === void 0 ? void 0 : props.renderMessageInput,
    renderTypingIndicator: props === null || props === void 0 ? void 0 : props.renderTypingIndicator,
    renderCustomSeparator: props === null || props === void 0 ? void 0 : props.renderCustomSeparator,
    renderFileUploadIcon: props === null || props === void 0 ? void 0 : props.renderFileUploadIcon,
    renderVoiceMessageIcon: props === null || props === void 0 ? void 0 : props.renderVoiceMessageIcon,
    renderSendMessageIcon: props === null || props === void 0 ? void 0 : props.renderSendMessageIcon
  }));
};

module.exports = Channel;
