// @sendbird/uikit-react/cjs/CreateOpenChannel.js
'use strict';

var React = require('react');
var CreateOpenChannel_components_CreateOpenChannelUI = require('./CreateOpenChannel/components/CreateOpenChannelUI.js');
var CreateOpenChannel_context = require('./CreateOpenChannel/context.js');
require('./LocalizationContext-58b6ed7c.js');
require('./stringSet-1539ac47.js');
require('./index-e8928da6.js');
require('./ui/Avatar.js');
require('./tslib.es6-c74b513f.js');
require('./ui/ImageRenderer.js');
require('./ui/Icon.js');
require('prop-types');
require('./uuid-9a117e16.js');
require('./index-b5a9b4b7.js');
require('./types-d0d770c3.js');
require('./ui/Modal.js');
require('react-dom');
require('./utils-eb6d90e0.js');
require('./index-54279ec2.js');
require('./index-b8ba4548.js');
require('./MediaQueryContext-6141d955.js');
require('./ui/IconButton.js');
require('./ui/Button.js');
require('./ui/Input.js');
require('./ui/TextButton.js');
require('./color-5bf5a8f5.js');
require('./useSendbirdStateContext.js');
require('./withSendbird.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function CreateOpenChannel(_a) {
  var className = _a.className,
    onCreateChannel = _a.onCreateChannel,
    onBeforeCreateChannel = _a.onBeforeCreateChannel,
    closeModal = _a.closeModal,
    renderHeader = _a.renderHeader,
    renderProfileInput = _a.renderProfileInput;
  return /*#__PURE__*/React__default["default"].createElement(CreateOpenChannel_context.CreateOpenChannelProvider, {
    className: className,
    onCreateChannel: onCreateChannel,
    onBeforeCreateChannel: onBeforeCreateChannel
  }, /*#__PURE__*/React__default["default"].createElement(CreateOpenChannel_components_CreateOpenChannelUI, {
    closeModal: closeModal,
    renderHeader: renderHeader,
    renderProfileInput: renderProfileInput
  }));
}

module.exports = CreateOpenChannel;
