// @sendbird/uikit-react/cjs/index-ce07a2a3.js
'use strict';

var React = require('react');
var index = require('./index-54279ec2.js');
var tokenize = require('./tokenize-d5286fa8.js');
var Message_context = require('./Message/context.js');
var ui_MentionLabel = require('./ui/MentionLabel.js');
var ui_LinkLabel = require('./ui/LinkLabel.js');
var ui_Label = require('./index-b5a9b4b7.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

// this function is used to generate a unique key for token in message
// it changes with updated time and index
// messageUpdatedAt is the key part of this key generator
function keyGenerator(createdAt, messageUpdatedAt, index) {
  return "sb-msg_".concat(createdAt, "_").concat(messageUpdatedAt, "_").concat(index);
}

function TextFragment(_a) {
  var tokens = _a.tokens;
  var messageStore = Message_context.useMessageContext();
  var message = messageStore === null || messageStore === void 0 ? void 0 : messageStore.message;
  var isByMe = messageStore === null || messageStore === void 0 ? void 0 : messageStore.isByMe;
  var updatedAt = message.updatedAt,
    createdAt = message.createdAt;
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, tokens === null || tokens === void 0 ? void 0 : tokens.map(function (token, idx) {
    var key = keyGenerator(createdAt, updatedAt, idx);
    return index.K(token.type).with(tokenize.TOKEN_TYPES.mention, function () {
      return /*#__PURE__*/React__default["default"].createElement("span", {
        className: "sendbird-word",
        key: key
      }, /*#__PURE__*/React__default["default"].createElement(ui_MentionLabel, {
        mentionTemplate: tokenize.USER_MENTION_PREFIX,
        mentionedUserId: token.userId,
        mentionedUserNickname: token.value,
        isByMe: isByMe
      }));
    }).with(tokenize.TOKEN_TYPES.url, function () {
      return /*#__PURE__*/React__default["default"].createElement("span", {
        className: "sendbird-word",
        key: key
      }, /*#__PURE__*/React__default["default"].createElement(ui_LinkLabel["default"], {
        className: "sendbird-word__url",
        src: token.value,
        type: ui_Label.LabelTypography.BODY_1
      }, token.value));
    }).otherwise(function () {
      return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, {
        key: key
      }, tokenize.getWhiteSpacePreservedText(token.value));
    });
  }));
}

exports.TextFragment = TextFragment;
