// @sendbird/uikit-react/cjs/OpenChannelList.js
'use strict';

var React = require('react');
var OpenChannelList_components_OpenChannelListUI = require('./OpenChannelList/components/OpenChannelListUI.js');
var OpenChannelList_context = require('./OpenChannelListProvider-a8560978.js');
require('./OpenChannelList/components/OpenChannelPreview.js');
require('./ui/Avatar.js');
require('./tslib.es6-c74b513f.js');
require('./ui/ImageRenderer.js');
require('./ui/Icon.js');
require('prop-types');
require('./uuid-9a117e16.js');
require('./index-b5a9b4b7.js');
require('./stringSet-1539ac47.js');
require('./ui/PlaceHolder.js');
require('./LocalizationContext-58b6ed7c.js');
require('./index-e8928da6.js');
require('./ui/Loader.js');
require('./ui/IconButton.js');
require('./CreateOpenChannel.js');
require('./CreateOpenChannel/components/CreateOpenChannelUI.js');
require('./types-d0d770c3.js');
require('./ui/Modal.js');
require('react-dom');
require('./utils-eb6d90e0.js');
require('./index-54279ec2.js');
require('./index-b8ba4548.js');
require('./MediaQueryContext-6141d955.js');
require('./ui/Button.js');
require('./ui/Input.js');
require('./ui/TextButton.js');
require('./color-5bf5a8f5.js');
require('./CreateOpenChannel/context.js');
require('./useSendbirdStateContext.js');
require('./withSendbird.js');
require('./consts-f54b15c2.js');
require('./pubSub/topics.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function OpenChannelList(_a) {
  var
    // provider
    className = _a.className,
    queries = _a.queries,
    onChannelSelected = _a.onChannelSelected,
    // ui
    renderHeader = _a.renderHeader,
    renderChannelPreview = _a.renderChannelPreview,
    renderPlaceHolderEmpty = _a.renderPlaceHolderEmpty,
    renderPlaceHolderError = _a.renderPlaceHolderError,
    renderPlaceHolderLoading = _a.renderPlaceHolderLoading;
  return /*#__PURE__*/React__default["default"].createElement(OpenChannelList_context.OpenChannelListProvider, {
    className: className,
    queries: queries,
    onChannelSelected: onChannelSelected
  }, /*#__PURE__*/React__default["default"].createElement(OpenChannelList_components_OpenChannelListUI, {
    renderHeader: renderHeader,
    renderChannelPreview: renderChannelPreview,
    renderPlaceHolderEmpty: renderPlaceHolderEmpty,
    renderPlaceHolderError: renderPlaceHolderError,
    renderPlaceHolderLoading: renderPlaceHolderLoading
  }));
}

module.exports = OpenChannelList;
