// @sendbird/uikit-react/cjs/RemoveMessageModal-afc3e268.js
'use strict';

var React = require('react');
var ui_Modal = require('./ui/Modal.js');
require('./tslib.es6-c74b513f.js');
require('./index-b5a9b4b7.js');
var types = require('./types-d0d770c3.js');
var LocalizationContext = require('./LocalizationContext-58b6ed7c.js');
var Thread_context = require('./Thread/context.js');
var stringFormatterUtils = require('./stringFormatterUtils-92dbef3c.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var RemoveMessage = function (props) {
  var _a;
  var onCancel = props.onCancel,
    onSubmit = props.onSubmit,
    message = props.message;
  var stringSet = React.useContext(LocalizationContext.LocalizationContext).stringSet;
  var deleteMessage = Thread_context.useThreadContext().deleteMessage;
  return /*#__PURE__*/React__default["default"].createElement(ui_Modal.Modal, {
    type: types.ButtonTypes.DANGER,
    disabled: ((_a = message === null || message === void 0 ? void 0 : message.threadInfo) === null || _a === void 0 ? void 0 : _a.replyCount) > 0,
    onCancel: onCancel,
    onSubmit: function () {
      deleteMessage(message).then(function () {
        onCancel === null || onCancel === void 0 ? void 0 : onCancel();
        onSubmit === null || onSubmit === void 0 ? void 0 : onSubmit();
      });
    },
    submitText: stringSet.MESSAGE_MENU__DELETE,
    titleText: stringFormatterUtils.getModalDeleteMessageTitle(stringSet, message)
  });
};

exports.RemoveMessage = RemoveMessage;
