// @sendbird/uikit-react/cjs/UserProfileContext-c6d7abb7.js
'use strict';

var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

/**
 * user profile goes deep inside the component tree
 * use this context as a short circuit to send in values
 */
var UserProfileContext = /*#__PURE__*/React__default["default"].createContext({
  disableUserProfile: true,
  isOpenChannel: false,
  renderUserProfile: null,
  onUserProfileMessage: null
});
var UserProfileProvider = function (_a) {
  var _b = _a.isOpenChannel,
    isOpenChannel = _b === void 0 ? false : _b,
    _c = _a.disableUserProfile,
    disableUserProfile = _c === void 0 ? false : _c,
    _d = _a.renderUserProfile,
    renderUserProfile = _d === void 0 ? null : _d,
    _e = _a.onUserProfileMessage,
    onUserProfileMessage = _e === void 0 ? null : _e,
    children = _a.children;
  return /*#__PURE__*/React__default["default"].createElement(UserProfileContext.Provider, {
    value: {
      isOpenChannel: isOpenChannel,
      disableUserProfile: disableUserProfile,
      renderUserProfile: renderUserProfile,
      onUserProfileMessage: onUserProfileMessage
    }
  }, children);
};

exports.UserProfileContext = UserProfileContext;
exports.UserProfileProvider = UserProfileProvider;
