// @sendbird/uikit-react/ChannelList/components/AddChannel.js
import React__default, { useState } from 'react';
import IconButton from '../../ui/IconButton.js';
import Icon, { IconTypes, IconColors } from '../../ui/Icon.js';
import CreateChannel from '../../CreateChannel.js';
import { useSendbirdStateContext } from '../../useSendbirdStateContext.js';
import { u as useChannelListContext } from '../../ChannelListProvider-eccd99c6.js';
import '../../tslib.es6-c2fb729e.js';
import 'prop-types';
import '../../CreateChannel/components/CreateChannelUI.js';
import '../../CreateChannelProvider-a2885972.js';
import '../../sendbirdSelectors.js';
import '../../pubSub/topics.js';
import '../../utils-6c50a6d1.js';
import '../../CreateChannel/components/InviteUsers.js';
import '../../LocalizationContext-287ee149.js';
import '../../stringSet-6422c6b9.js';
import '../../index-e4277f9c.js';
import '../../ui/Modal.js';
import 'react-dom';
import '../../index-09847c0f.js';
import '../../index-04bd1a59.js';
import '../../MediaQueryContext-114ca049.js';
import '../../ui/Button.js';
import '../../index-92741c7e.js';
import '../../types-9d43ac6e.js';
import '../../ui/UserListItem.js';
import '../../UserProfileContext-2001a468.js';
import '../../ui/Avatar.js';
import '../../ui/ImageRenderer.js';
import '../../uuid-c15b8b3b.js';
import '../../ui/MutedAvatarOverlay.js';
import '../../ui/Checkbox.js';
import '../../ui/UserProfile.js';
import '../../ui/ContextMenu.js';
import '../../ui/SortByRow.js';
import '../../index-776554c0.js';
import '../../utils/message/getOutgoingMessageState.js';
import '../../CreateChannel/components/SelectChannelType.js';
import '../../withSendbird.js';
import '@sendbird/chat/groupChannel';
import '../../consts-d9c5f25a.js';

var AddChannel = function () {
  var _a;
  var _b = useState(false),
    showModal = _b[0],
    setShowModal = _b[1];
  var state = useSendbirdStateContext();
  var isOnline = (_a = state === null || state === void 0 ? void 0 : state.config) === null || _a === void 0 ? void 0 : _a.isOnline;
  var disabled = !isOnline;
  var _c = useChannelListContext(),
    overrideInviteUser = _c.overrideInviteUser,
    onBeforeCreateChannel = _c.onBeforeCreateChannel;
  return /*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement(IconButton, {
    height: "32px",
    width: "32px",
    onClick: function () {
      setShowModal(true);
    },
    disabled: disabled
  }, /*#__PURE__*/React__default.createElement(Icon, {
    type: IconTypes.CREATE,
    fillColor: IconColors.PRIMARY,
    width: "24px",
    height: "24px"
  })), showModal && /*#__PURE__*/React__default.createElement(CreateChannel, {
    onCancel: function () {
      setShowModal(false);
    },
    overrideInviteUser: overrideInviteUser,
    onCreateChannel: function () {
      setShowModal(false);
    },
    onBeforeCreateChannel: onBeforeCreateChannel
  }));
};

export { AddChannel, AddChannel as default };
