// @sendbird/uikit-react/cjs/OpenChannel/components/OpenChannelInput.js
'use strict';

var React = require('react');
var LocalizationContext = require('../../LocalizationContext-58b6ed7c.js');
var ui_MessageInput = require('../../ui/MessageInput.js');
var OpenChannel_context = require('../../OpenChannelProvider-39f13647.js');
require('../../stringSet-1539ac47.js');
require('../../index-e8928da6.js');
require('../../_rollupPluginBabelHelpers-f5b0dfb9.js');
require('prop-types');
require('../../const-876980e8.js');
require('../../const-c9e7a340.js');
require('../../ui/IconButton.js');
require('../../tslib.es6-c74b513f.js');
require('../../ui/Button.js');
require('../../index-b5a9b4b7.js');
require('../../types-d0d770c3.js');
require('../../index-cc2e05a4.js');
require('dompurify');
require('../../consts-b3c4f548.js');
require('../../consts-61d83828.js');
require('../../consts-b711321c.js');
require('../../ui/Icon.js');
require('../../useSendbirdStateContext.js');
require('../../withSendbird.js');
require('../../index-dbe8dd84.js');
require('../../utils/message/getOutgoingMessageState.js');
require('../../index-54279ec2.js');
require('../../tokenize-d5286fa8.js');
require('../../index-3448df64.js');
require('../../UserProfileContext-c6d7abb7.js');
require('../../compareIds-28c6af62.js');
require('../../pubSub/topics.js');
require('@sendbird/chat');
require('@sendbird/chat/openChannel');
require('../../uuid-9a117e16.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var MessageInputWrapper = function (props, ref) {
  var _a = OpenChannel_context.useOpenChannelContext(),
    currentOpenChannel = _a.currentOpenChannel,
    disabled = _a.disabled,
    handleSendMessage = _a.handleSendMessage,
    handleFileUpload = _a.handleFileUpload,
    amIMuted = _a.amIMuted;
  var channel = currentOpenChannel;
  var stringSet = React.useContext(LocalizationContext.LocalizationContext).stringSet;
  var value = props.value;
  function getPlaceHolderString() {
    if (amIMuted) {
      return stringSet.MESSAGE_INPUT__PLACE_HOLDER__MUTED;
    }
    if (disabled) {
      return stringSet.MESSAGE_INPUT__PLACE_HOLDER__DISABLED;
    }
    return '';
  }
  if (!channel) {
    return null;
  }
  return /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-openchannel-footer"
  }, /*#__PURE__*/React__default["default"].createElement(ui_MessageInput, {
    channel: currentOpenChannel,
    ref: ref,
    value: value,
    disabled: disabled,
    isVoiceMessageEnabled: false,
    onSendMessage: handleSendMessage,
    onFileUpload: handleFileUpload,
    placeholder: getPlaceHolderString()
  }));
};
var OpenChannelInput = /*#__PURE__*/React__default["default"].forwardRef(MessageInputWrapper);

module.exports = OpenChannelInput;
