// @sendbird/uikit-react/ui/ReactionButton.js
import { _ as _extends } from '../_rollupPluginBabelHelpers-e92eb032.js';
import { c as __spreadArray } from '../tslib.es6-c2fb729e.js';
import React__default from 'react';
import { u as useLongPress } from '../useLongPress-f8655a14.js';
import { n as noop } from '../utils-6c50a6d1.js';
import '../MediaQueryContext-114ca049.js';

var ReactionButton = /*#__PURE__*/React__default.forwardRef(function (props, ref) {
  var className = props.className,
    width = props.width,
    height = props.height,
    selected = props.selected,
    _a = props.dataSbId,
    dataSbId = _a === void 0 ? '' : _a,
    onClick = props.onClick,
    children = props.children;
  var onClickHandler = useLongPress({
    onLongPress: noop,
    onClick: onClick
  }, {
    shouldPreventDefault: true,
    shouldStopPropagation: true
  });
  return /*#__PURE__*/React__default.createElement("div", _extends({
    className: __spreadArray(__spreadArray([], Array.isArray(className) ? className : [className], true), ["sendbird-reaction-button".concat(selected ? '--selected' : '')], false).join(' '),
    ref: ref,
    role: "button",
    style: {
      width: width,
      height: height
    }
  }, onClickHandler, {
    tabIndex: 0,
    "data-sb-id": dataSbId
  }), /*#__PURE__*/React__default.createElement("div", {
    className: "sendbird-reaction-button__inner"
  }, children));
});

export { ReactionButton as default };
