// @sendbird/uikit-react/ui/ThumbnailMessageItemBody.js
import { _ as _extends } from '../_rollupPluginBabelHelpers-e92eb032.js';
import React__default, { useState } from 'react';
import Icon, { IconTypes, IconColors } from './Icon.js';
import ImageRenderer from './ImageRenderer.js';
import { H as isSentMessage, o as getClassName, t as isVideoMessage, x as isGifMessage } from '../index-776554c0.js';
import { n as noop } from '../utils-6c50a6d1.js';
import { u as useLongPress } from '../useLongPress-f8655a14.js';
import { b as getMessageFirstFileThumbnailUrl, g as getMessageFirstFileUrl, c as getMessageFirstFileType } from '../utils-bfc0ab3b.js';
import 'prop-types';
import '../tslib.es6-c2fb729e.js';
import '../utils/message/getOutgoingMessageState.js';
import '../index-09847c0f.js';
import '../MediaQueryContext-114ca049.js';

function ThumbnailMessageItemBody(_a) {
  var _b, _c;
  var _d = _a.className,
    className = _d === void 0 ? '' : _d,
    message = _a.message,
    _e = _a.isByMe,
    isByMe = _e === void 0 ? false : _e,
    _f = _a.mouseHover,
    mouseHover = _f === void 0 ? false : _f,
    _g = _a.isReactionEnabled,
    isReactionEnabled = _g === void 0 ? false : _g,
    _h = _a.showFileViewer,
    showFileViewer = _h === void 0 ? noop : _h,
    _j = _a.style,
    style = _j === void 0 ? {} : _j;
  var thumbnailUrl = getMessageFirstFileThumbnailUrl(message);
  var _k = useState(false),
    imageRendered = _k[0],
    setImageRendered = _k[1];
  var onClickHandler = useLongPress({
    onLongPress: noop,
    onClick: function () {
      if (isSentMessage(message)) {
        showFileViewer === null || showFileViewer === void 0 ? void 0 : showFileViewer(true);
      }
    }
  });
  return /*#__PURE__*/React__default.createElement("div", _extends({
    className: getClassName([className, 'sendbird-thumbnail-message-item-body', isByMe ? 'outgoing' : 'incoming', mouseHover ? 'mouse-hover' : '', isReactionEnabled && ((_c = (_b = message.reactions) === null || _b === void 0 ? void 0 : _b.length) !== null && _c !== void 0 ? _c : 0) > 0 ? 'reactions' : ''])
  }, onClickHandler), /*#__PURE__*/React__default.createElement(ImageRenderer, {
    className: "sendbird-thumbnail-message-item-body__thumbnail",
    url: thumbnailUrl || getMessageFirstFileUrl(message),
    alt: getMessageFirstFileType(message),
    width: (style === null || style === void 0 ? void 0 : style.width) || '360px',
    height: (style === null || style === void 0 ? void 0 : style.height) || '270px',
    onLoad: function () {
      setImageRendered(true);
    },
    placeHolder: function (style_) {
      return /*#__PURE__*/React__default.createElement("div", {
        className: "sendbird-thumbnail-message-item-body__placeholder",
        style: style_
      });
    }
  }), isVideoMessage(message) && !thumbnailUrl && !imageRendered && /*#__PURE__*/React__default.createElement("video", {
    className: "sendbird-thumbnail-message-item-body__video"
  }, /*#__PURE__*/React__default.createElement("source", {
    src: getMessageFirstFileUrl(message),
    type: getMessageFirstFileType(message)
  })), /*#__PURE__*/React__default.createElement("div", {
    className: "sendbird-thumbnail-message-item-body__image-cover"
  }), (isVideoMessage(message) || isGifMessage(message)) && /*#__PURE__*/React__default.createElement("div", {
    className: "sendbird-thumbnail-message-item-body__icon-wrapper"
  }, /*#__PURE__*/React__default.createElement("div", {
    className: "sendbird-thumbnail-message-item-body__icon-wrapper__icon"
  }, /*#__PURE__*/React__default.createElement(Icon, {
    type: isVideoMessage(message) ? IconTypes.PLAY : IconTypes.GIF,
    fillColor: IconColors.THUMBNAIL_ICON,
    width: "34px",
    height: "34px"
  }))));
}

export { ThumbnailMessageItemBody as default };
