// @sendbird/uikit-react/cjs/ChannelList/components/ChannelListHeader.js
'use strict';

var React = require('react');
var LocalizationContext = require('../../LocalizationContext-58b6ed7c.js');
var useSendbirdStateContext = require('../../useSendbirdStateContext.js');
var ui_Label = require('../../index-b5a9b4b7.js');
var ui_Avatar = require('../../ui/Avatar.js');
require('../../stringSet-1539ac47.js');
require('../../index-e8928da6.js');
require('../../withSendbird.js');
require('../../tslib.es6-c74b513f.js');
require('prop-types');
require('../../ui/ImageRenderer.js');
require('../../ui/Icon.js');
require('../../uuid-9a117e16.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var ChannelListHeader = function (_a) {
  var renderHeader = _a.renderHeader,
    renderTitle = _a.renderTitle,
    renderIconButton = _a.renderIconButton,
    onEdit = _a.onEdit,
    allowProfileEdit = _a.allowProfileEdit;
  var _b = useSendbirdStateContext.useSendbirdStateContext(),
    stores = _b.stores,
    config = _b.config;
  var user = stores.userStore.user;
  var logger = config.logger;
  var stringSet = React.useContext(LocalizationContext.LocalizationContext).stringSet;
  if (renderHeader) {
    logger === null || logger === void 0 ? void 0 : logger.warning('Recomend to use "renderTitle" instead of "renderHeader". It will be deprecated.');
  }
  // renderTitle should have higher priority
  var titleRenderer = renderHeader || renderTitle;
  return /*#__PURE__*/React__default["default"].createElement("div", {
    className: ['sendbird-channel-header', allowProfileEdit ? 'sendbird-channel-header--allow-edit' : ''].join(' ')
  }, (titleRenderer === null || titleRenderer === void 0 ? void 0 : titleRenderer()) || /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-channel-header__title",
    role: "button",
    onClick: function () {
      onEdit === null || onEdit === void 0 ? void 0 : onEdit();
    },
    onKeyDown: function () {
      onEdit === null || onEdit === void 0 ? void 0 : onEdit();
    },
    tabIndex: 0
  }, /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-channel-header__title__left"
  }, /*#__PURE__*/React__default["default"].createElement(ui_Avatar["default"], {
    width: "32px",
    height: "32px",
    src: user.profileUrl,
    alt: user.nickname
  })), /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-channel-header__title__right"
  }, /*#__PURE__*/React__default["default"].createElement(ui_Label.Label, {
    className: "sendbird-channel-header__title__right__name",
    type: ui_Label.LabelTypography.SUBTITLE_2,
    color: ui_Label.LabelColors.ONBACKGROUND_1
  }, user.nickname || stringSet.NO_NAME), /*#__PURE__*/React__default["default"].createElement(ui_Label.Label, {
    className: "sendbird-channel-header__title__right__user-id",
    type: ui_Label.LabelTypography.BODY_2,
    color: ui_Label.LabelColors.ONBACKGROUND_2
  }, user.userId))), /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-channel-header__right-icon"
  }, renderIconButton === null || renderIconButton === void 0 ? void 0 : renderIconButton()));
};

module.exports = ChannelListHeader;
