// @sendbird/uikit-react/Channel/components/RemoveMessageModal.js
import React__default, { useContext } from 'react';
import { Modal } from '../../ui/Modal.js';
import '../../tslib.es6-c2fb729e.js';
import '../../index-92741c7e.js';
import { B as ButtonTypes } from '../../types-9d43ac6e.js';
import { a as LocalizationContext } from '../../LocalizationContext-287ee149.js';
import { u as useChannelContext } from '../../ChannelProvider-9ba5893a.js';
import { g as getModalDeleteMessageTitle } from '../../stringFormatterUtils-abbe70ba.js';
import 'react-dom';
import '../../utils-6c50a6d1.js';
import '../../index-09847c0f.js';
import '../../index-04bd1a59.js';
import '../../MediaQueryContext-114ca049.js';
import '../../ui/IconButton.js';
import '../../ui/Button.js';
import 'prop-types';
import '../../stringSet-6422c6b9.js';
import '../../ui/Icon.js';
import '../../index-e4277f9c.js';
import '../../UserProfileContext-2001a468.js';
import '../../useSendbirdStateContext.js';
import '../../withSendbird.js';
import '../../const-73db0277.js';
import '../../utils-aa4086dc.js';
import '../../index-2d27aeef.js';
import '../../index-776554c0.js';
import '../../utils/message/getOutgoingMessageState.js';
import '@sendbird/chat/message';
import '../../useSendMultipleFilesMessage-b8256698.js';
import '../../pubSub/topics.js';
import '../../compareIds-7e517404.js';
import '@sendbird/chat/groupChannel';
import '../../uuid-c15b8b3b.js';
import '../../_rollupPluginBabelHelpers-e92eb032.js';
import '../../consts-d9c5f25a.js';
import '../../resolvedReplyType-fb987595.js';

var RemoveMessage = function (props) {
  var _a;
  var _b = props.onSubmit,
    onSubmit = _b === void 0 ? function () {} : _b,
    onCancel = props.onCancel,
    message = props.message;
  var stringSet = useContext(LocalizationContext).stringSet;
  var deleteMessage = useChannelContext().deleteMessage;
  return /*#__PURE__*/React__default.createElement(Modal, {
    type: ButtonTypes.DANGER,
    disabled: ((_a = message === null || message === void 0 ? void 0 : message.threadInfo) === null || _a === void 0 ? void 0 : _a.replyCount) > 0,
    onCancel: onCancel,
    onSubmit: function () {
      deleteMessage(message).then(function () {
        // For other status such as PENDING, SCHEDULED, and CANCELED,
        // invalid parameters error is thrown so nothing happens.
        onSubmit();
        onCancel();
      });
    },
    submitText: stringSet.MESSAGE_MENU__DELETE,
    titleText: getModalDeleteMessageTitle(stringSet, message)
  });
};

export { RemoveMessage as default };
